Code/Framework/AzToolsFramework/AzToolsFramework/AssetBrowser/AssetBrowserThumbnailViewProxyModel.cpp

```diff
@@ -38,18 +46,93 @@ namespace AzToolsFramework
             {
             case Qt::DecorationRole:
                 {
-                    SharedThumbnail thumbnail;
-                    QPersistentModelIndex persistentIndex;
-
-                    ThumbnailerRequestBus::BroadcastResult(
-                        thumbnail, &ThumbnailerRequests::GetThumbnail, assetBrowserEntry->GetThumbnailKey());
-                    AZ_Assert(thumbnail, "The shared thumbnail was not available from the ThumbnailerRequestBus.");
-                    if (!thumbnail || thumbnail->GetState() == Thumbnail::State::Failed)
+                    // If this entry has a custom previewer, use its thumbnail
+                    AZ::EBusAggregateResults<const PreviewerFactory*> factories;
+                    PreviewerRequestBus::BroadcastResult(factories, &PreviewerRequests::GetPreviewerFactory, assetBrowserEntry);
+                    for (const auto factory : factories.values)
                     {
-                        return QIcon{};
+                        if (factory)
+                        {
+                            SharedThumbnail thumbnail;
+
+                            ThumbnailerRequestBus::BroadcastResult(
+                                thumbnail, &ThumbnailerRequests::GetThumbnail, assetBrowserEntry->GetThumbnailKey());
+                            AZ_Assert(thumbnail, "The shared thumbnail was not available from the ThumbnailerRequestBus.");
+                            if (thumbnail && thumbnail->GetState() != Thumbnail::State::Failed)
+                            {
+                                return thumbnail->GetPixmap();
+                            }
+                        }
                     }
```

Comment:
Why does this need to enumerate the factories if it doesn't directly do anything with them? Couldn't it just be this?'

                            SharedThumbnail thumbnail;

                            ThumbnailerRequestBus::BroadcastResult(
                                thumbnail, &ThumbnailerRequests::GetThumbnail, assetBrowserEntry->GetThumbnailKey());
                            AZ_Assert(thumbnail, "The shared thumbnail was not available from the ThumbnailerRequestBus.");
                            if (thumbnail && thumbnail->GetState() != Thumbnail::State::Failed)
                            {
                                return thumbnail->GetPixmap();
                            }

Reply:
So I ran into an issue where if I simply tried to get all thumbnails, the resulting icons were tiny. Scaling them up resulted in extremely blurry icons, and if I smoothed them then they would be incredibly blurry. However, this is only an issue for entries that do not have a previewer. Therefore, I'm enumerating the factories as a way to check for entries with a previewer, because we can just return the thumbnail for that set of entries.

The code below this part digs through the directories and returns a path to each icon file so that we can create an appropriately sized icon for that entry. I'm not sure how else to differentiate between these two sets of entries, so I'm open to any suggestions :)

Reply:
OK. Since you're not using the factories, could you just check if the container is empty instead of looping over them? Very supportive of any optimization, simplification, improvement to the thumbnail system. Work still needs to be done to fix all of the loading animations from getting stuck or occupying all of the available threads. Separate issue though.

Reply:
Unfortunately the container is never empty and always contains two factories, which are nullptrs most of the time. I tried seeing if there was a way I could query the factory for the thumbnail, but I'm not seeing that info stored anywhere 😞